### ISSUE TITLE
node_modules recompile: ‘legacy’ not recompiling js

### ISSUE THREAD

**Issue by mfen** [meteor/meteor#11215]
I was trying to recompile `meteor-accounts-t9n` for legacy, as its usage of `class` breaks our app in ie11, but it's not recompiled in the legacy bundle. Changing the recompile option to `modern`, or symlinking the module in `/imports`, does trigger recompilation.

Meteor v1.11.1 
OS X v10.15.5

**package.json:**
```json
"meteor": {
    "mainModule": {
      "client": "client/main.js",
      "server": "server/main.js"
    },
    "nodeModules": {
      "recompile": {
        "meteor-accounts-t9n": "legacy"
      }
    }
  },
```

Using a checkout of meteor, it looks like package-source’s `_findSources` pass for web.browser is first discovering them with `nodeModulesReadOptions` excluding `*.js(on)` and caching node_modules's sources in `_findSourcesCache`. Then the web.browser.legacy `_findSources` pass hits the `_findSourcesCache` where .js was excluded, and so doesn’t add the sources to be recompiled for the legacy unibuild.

Oddly, I see claims of recompile for legacy only [working for others](https://github.com/meteor/meteor/pull/10603#issuecomment-559319720
).

**Comment by ArchiCroc** [comment]
I'm seeing this too.  My app uses the npm query-string module and recompile legacy works on 1.10.2 but not on 1.11.1 or the new 2.0-beta.3. Not sure how to best troubleshoot this further. You can pull my boilerplate app https://github.com/ArchiCroc/pup-plus and try it for your self.

**Comment by hmelenok** [comment]
Same situation. Previously working `package.json` config:
```json
"meteor": {
    "nodeModules": {
      "recompile": {
        "email-regex": [
          "legacy",
          "modern"
        ],
        "debug": [
          "legacy",
          "modern"
        ],
        "normalize-url": [
          "legacy",
          "modern"
        ],
        "serialize-error": [
          "legacy"
        ],
        "isemail": [
          "legacy"
        ],
        "punycode": [
          "legacy"
        ],
```

**Comment by djhutch** [comment]
Same here.
`    "meteor": {
        "nodeModules": {
            "recompile": {
                "query-string": "legacy",
                "rotating-file-stream": "legacy",
                "split-on-first": "legacy",
                "strict-uri-encode": "legacy",
                "debug": "legacy",
                "xml-formatter": "legacy",
                "xml-parser-xo": "legacy",
                "react-table-sticky": "legacy"
            }
        }
    },
`

**Comment by nathan-muir** [comment]
Can confirm that `meteor.nodeModules.recompile` `legacy` works with meteor tool 1.10.2, but not with 1.11.

**Comment by nathan-muir** [comment]
Output from `git bisect` -- 

```
# bad: [3c96818a18fe88193841a73d5fa297e14bdefe21] Set 1.11 release date on History
# good: [23dbf94bf74cfcfb2c017764d7306bb52e5eef4c] Bump package versions for 1.10.2 release
git bisect start '3c96818a' '23dbf94b'
# good: [3181a2c698383416fe423e55859e4fc39baaf675] Merge branch 'devel' into release-1.10.3
git bisect good 3181a2c698383416fe423e55859e4fc39baaf675
# skip: [ce6b4832b425bc3b3628c59af4dacd69c7416896] Update npm to 6.14.6
git bisect skip ce6b4832b425bc3b3628c59af4dacd69c7416896
# good: [a9983a2aadd0c00416fabf77b2d18e1fb84022d1] Bump minifier-css package version to 1.5.2
git bisect good a9983a2aadd0c00416fabf77b2d18e1fb84022d1
# bad: [ca8a6285e6a7534ac2ffc4956ac16c9515cffc45] Bump package versions for 1.10.3-beta.3 release
git bisect bad ca8a6285e6a7534ac2ffc4956ac16c9515cffc45
# bad: [3a736233b415a531ef504e898a3a9a31eecc4cb0] Merge remote-tracking branch 'origin/devel' into find-node_module-performance
git bisect bad 3a736233b415a531ef504e898a3a9a31eecc4cb0
# bad: [18629d25ae4091c3dfe889e33c456fa2df71b161] Show arch name in profile for _findSources
git bisect bad 18629d25ae4091c3dfe889e33c456fa2df71b161
# bad: [93557870409c905736cae2f9940033363e2c0331] Do not watch node_modules in _findSources
git bisect bad 93557870409c905736cae2f9940033363e2c0331
# bad: [e763d9598eb4192f5dda0616207a8e0594f59cfa] Allow cache keys to be shared between archs
git bisect bad e763d9598eb4192f5dda0616207a8e0594f59cfa
# first bad commit: [e763d9598eb4192f5dda0616207a8e0594f59cfa] Allow cache keys to be shared between archs
```

Seems the culprit is `e763d9598eb4192f5dda0616207a8e0594f59cfa`. Reverting this commit atop 1.11 fixes the recompilation of npm modules for `legacy` browsers.

**Comment by HemalR** [comment]
Is there a workaround/fix for this or is the only solution to roll back to 1.10 ? 

(Damn you IE!!!)

**Comment by mfen** [comment]
> Is there a workaround/fix for this or is the only solution to roll back to 1.10 ?

@HemalR they're not great, but some workarounds from the opening comment:
> Changing the recompile option to modern, or symlinking the module in /imports, does trigger recompilation.

**Comment by filipenevola** [comment]
Hi @HemalR and @mfen I believe we could change the cache key and release this fix in the next version of Meteor. See https://github.com/meteor/meteor/pull/11247#issuecomment-755436197

**Comment by HemalR** [comment]
That would be great! I'll keep an ear out - for the time being we have had
to revert to the older version because we have to support IE11 :/

**Comment by filipenevola** [comment]
Hi @HemalR and @mfen, could you try Meteor 2.0-rc.0?

Read how here https://github.com/meteor/meteor/pull/11206#issuecomment-761872182

**Comment by HemalR** [comment]
Looks good. I'll give it a crack and do some testing and report back.

**Comment by HemalR** [comment]
@filipenevola 

Update, apologies for the delay in getting to this:

I've tried the 2.0 build* and it doesn't seem to be fixed yet. It's breaking on an arrow function which is a part of a package that is under the recompile list:

```
"meteor": {
    "nodeModules": {
      "recompile": {
        "query-string": "legacy",
        "strict-uri-encode": "legacy",
        "simpl-schema": "legacy",
        "split-on-first": "legacy",
        "dequal": "legacy"
      }
    }
  },
```

![image](https://user-images.githubusercontent.com/15634303/106452087-5eb0a500-64c2-11eb-9c1f-a5b9b62af20e.png)

![image](https://user-images.githubusercontent.com/15634303/106452190-80aa2780-64c2-11eb-9cf1-e4cb4269231c.png)

Has anyone else here tested this successfully/confirming what I've found?

*not the rc because I think the 2.0 official was released before I tested this. Let me know if I have tested the wrong version.

**Comment by filipenevola** [comment]
Hi, could you add a log to my change to see if the cache key is being created correctly considering the modules to be recompiled?

**Comment by mfen** [comment]
Thanks for addressing this. I'd like to verify, but I've been a bit swamped recently. I should hopefully have some time to try out 2.0 and the fix later this month.

**Comment by HemalR** [comment]
@filipenevola 

> Hi, could you add a log to my change to see if the cache key is being created correctly considering the modules to be recompiled?

Is this directed at me? If so, I'm not entirely sure what/how to do this but happy to follow any instructions and give it a go.

**Comment by filipenevola** [comment]
Yes @HemalR, in this PR https://github.com/meteor/meteor/pull/11290/files I changed the baseCacheKey to include the modules to recompile.

I would like to be sure that in your case the cache key is correct.

You could run Meteor from a checkout (just git clone meteor and instead of running `meteor ...` you should run `<path to your local clone>/meteor ...`) and before that add some consoles after the cache key is created on Meteor code so you could see exactly what is the cache. Probably the cache key is missing some part that is important in your app to detect the cache as invalid.

Your console.log could go here https://github.com/meteor/meteor/blob/0111ff74004bb95acc1aad0010721862aa385e5a/tools/isobuild/package-source.js#L1275

`console.log('baseCacheKey', baseCacheKey);`

**Comment by HemalR** [comment]
Finally got round to trying to update this again and trying out the cache key stuff but it seems that it's already been fixed with 2.2 👍

I think we can close this.

Thank you very much @filipenevola
